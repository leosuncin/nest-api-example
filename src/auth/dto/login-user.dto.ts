import {
  IsDefined,
  IsNotEmpty,
  IsString,
  Matches,
  MaxLength,
  MinLength,
} from 'class-validator';

import { ValidateCredential } from '~auth/validators/validate-credential.validator';
import { Trim } from '~common/decorators/trim.decorator';

export class LoginUser {
  @IsDefined()
  @IsString()
  @IsNotEmpty()
  @MinLength(8)
  @MaxLength(30)
  @ValidateCredential()
  readonly password!: string;

  @Trim()
  @IsDefined()
  @IsString()
  @IsNotEmpty()
  @MaxLength(30)
  @Matches(/^[\w.-]+$/i)
  @ValidateCredential()
  readonly username!: string;
}
