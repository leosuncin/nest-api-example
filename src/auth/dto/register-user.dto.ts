import {
  IsDefined,
  IsEmail,
  IsNotEmpty,
  IsString,
  Matches,
  MaxLength,
  MinLength,
} from 'class-validator';

import { NormalizeEmail } from '~auth/decorators/normalize-email.decorator';
import { IsAlreadyRegister } from '~auth/validators/is-already-register.validator';
import { IsNotTheSame } from '~auth/validators/is-not-the-same';
import { IsNotVulnerable } from '~auth/validators/is-not-vulnerable.validator';
import { Trim } from '~common/decorators/trim.decorator';

export class RegisterUser {
  @Trim()
  @NormalizeEmail()
  @IsDefined()
  @IsString()
  @IsNotEmpty()
  @IsEmail()
  @IsAlreadyRegister()
  readonly email!: string;

  @IsDefined()
  @IsString()
  @IsNotEmpty()
  @MinLength(8)
  @MaxLength(30)
  @IsNotTheSame<RegisterUser>('username')
  @IsNotVulnerable()
  readonly password!: string;

  @Trim()
  @IsDefined()
  @IsString()
  @IsNotEmpty()
  @MaxLength(30)
  @Matches(/^[\w.-]+$/i)
  @IsAlreadyRegister()
  readonly username!: string;
}
