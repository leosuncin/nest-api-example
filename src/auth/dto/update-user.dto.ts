import {
  Allow,
  IsDefined,
  IsEmail,
  IsNotEmpty,
  IsOptional,
  IsString,
  IsUrl,
  Matches,
  MaxLength,
  MinLength,
  ValidateIf,
} from 'class-validator';

import { IsAlreadyRegister } from '~auth/validators/is-already-register.validator';
import { IsNotVulnerable } from '~auth/validators/is-not-vulnerable.validator';
import { ValidateCredential } from '~auth/validators/validate-credential.validator';

export class UpdateUser {
  @Allow()
  id?: string;

  @IsOptional()
  @IsString()
  @IsNotEmpty()
  @IsEmail()
  @IsAlreadyRegister()
  readonly email?: string;

  @IsOptional()
  @IsString()
  @IsNotEmpty()
  @MaxLength(30)
  @Matches(/^[\w.-]+$/i)
  @IsAlreadyRegister()
  readonly username?: string;

  @ValidateIf((object: UpdateUser) => typeof object.newPassword === 'string')
  @IsDefined()
  @IsString()
  @IsNotEmpty()
  @MinLength(8)
  @MaxLength(30)
  @ValidateCredential()
  readonly password?: string;

  @IsOptional()
  @IsString()
  @IsNotEmpty()
  @MinLength(8)
  @MaxLength(30)
  @IsNotVulnerable()
  readonly newPassword?: string;

  @IsOptional()
  @IsString()
  @IsUrl({
    allow_fragments: false,
    protocols: ['http', 'https'],
    require_protocol: true,
    require_valid_protocol: true,
  })
  readonly image?: string;

  @IsOptional()
  @IsString()
  @IsNotEmpty()
  readonly bio?: string;
}
