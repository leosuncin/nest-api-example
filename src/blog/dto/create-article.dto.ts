import { Allow, IsDefined, IsNotEmpty, IsString } from 'class-validator';

import { User } from '~auth/entities/user.entity';

export class CreateArticle {
  @IsDefined()
  @IsString()
  @IsNotEmpty()
  readonly title!: string;

  @IsDefined()
  @IsString()
  @IsNotEmpty()
  readonly content!: string;

  @Allow()
  readonly author!: User;
}
