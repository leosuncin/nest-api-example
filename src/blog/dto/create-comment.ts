import {
  Allow,
  IsDefined,
  IsNotEmpty,
  IsString,
  IsUUID,
} from 'class-validator';

import { User } from '~auth/entities/user.entity';
import { Article } from '~blog/entities/article.entity';
import { ArticleExist } from '~blog/validators/article-exist.validator';

export class CreateComment {
  @IsDefined()
  @IsString()
  @IsNotEmpty()
  readonly body!: string;

  @IsDefined()
  @IsUUID()
  @ArticleExist()
  readonly article!: Article;

  @Allow()
  readonly author!: User;
}
