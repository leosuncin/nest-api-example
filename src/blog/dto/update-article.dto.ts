import { OmitType, PartialType } from '@nestjs/mapped-types';

import { CreateArticle } from '~blog/dto/create-article.dto';

export class UpdateArticle extends PartialType(
  OmitType(CreateArticle, ['author']),
) {}
