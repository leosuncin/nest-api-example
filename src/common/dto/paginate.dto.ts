import { Transform } from 'class-transformer';
import { IsOptional, IsPositive, Max } from 'class-validator';

export class Paginate {
  @IsOptional()
  @Transform(({ value }) => Number.parseInt(value as string, 10))
  @IsPositive()
  @Max(100)
  readonly limit = 10;

  @IsOptional()
  @Transform(({ value }) => Number.parseInt(value as string, 10))
  @IsPositive()
  readonly page = 1;
}
