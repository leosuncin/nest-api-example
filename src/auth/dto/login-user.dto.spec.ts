import { faker } from '@faker-js/faker';
import { Test } from '@nestjs/testing';
import { plainToInstance } from 'class-transformer';
import { useContainer, validate } from 'class-validator';
import fc from 'fast-check';
import { createMockInstance } from 'jest-create-mock-instance';

import { LoginUser } from '~auth/dto/login-user.dto';
import { loginUserFactory } from '~auth/factories/login-user.factory';
import { login } from '~auth/fixtures/credentials';
import { AuthenticationService } from '~auth/services/authentication.service';
import { ValidateCredentialConstraint } from '~auth/validators/validate-credential.validator';

describe('Login user validations', () => {
  beforeAll(async () => {
    const module = await Test.createTestingModule({
      providers: [ValidateCredentialConstraint],
    })
      .useMocker((token) => {
        if (token === AuthenticationService) {
          const mock = createMockInstance(AuthenticationService);
          mock.verifyCredentials.mockResolvedValue(true);

          return mock;
        }

        return;
      })
      .compile();

    useContainer(module, { fallbackOnErrors: true });
  });

  it('should be validated', async () => {
    await fc.assert(
      fc.asyncProperty(
        fc
          .integer()
          .noBias()
          .noShrink()
          .map((seed) => {
            faker.seed(seed);

            return loginUserFactory.buildOne();
          }),
        async (data) => {
          const errors = await validate(data);

          expect(errors).toHaveLength(0);
        },
      ),
    );
  });

  it('should require all properties', async () => {
    await expect(validate(new LoginUser())).resolves.toMatchSnapshot();
  });

  it('should find the errors in the password', async () => {
    await fc.assert(
      fc.asyncProperty(
        fc
          .record(
            {
              password: fc.oneof(
                fc.nat(),
                fc.string({ maxLength: 7 }),
                fc.string({ minLength: 31 }),
              ),
              username: fc.constant(login.username),
            },
            { requiredKeys: ['username'] },
          )
          .map((plain) => plainToInstance(LoginUser, plain)),
        async (data) => {
          const errors = await validate(data);

          expect(errors).toHaveLength(1);
          expect(errors[0]).toHaveProperty('property', 'password');
        },
      ),
    );
  });

  it('should find the errors in the username', async () => {
    await fc.assert(
      fc.asyncProperty(
        fc
          .record(
            {
              password: fc.constant(login.password),
              username: fc.oneof(
                fc.nat(),
                fc.constant(''),
                fc.string({ minLength: 31 }),
              ),
            },
            { requiredKeys: ['password'] },
          )
          .map((plain) => plainToInstance(LoginUser, plain)),
        async (data) => {
          const errors = await validate(data);

          expect(errors).toHaveLength(1);
          expect(errors[0]).toHaveProperty('property', 'username');
        },
      ),
    );
  });
});
