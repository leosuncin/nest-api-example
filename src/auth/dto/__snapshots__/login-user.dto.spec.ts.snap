// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login user validations should require all properties 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isDefined": "password should not be null or undefined",
      "isNotEmpty": "password should not be empty",
      "isString": "password must be a string",
      "maxLength": "password must be shorter than or equal to 30 characters",
      "minLength": "password must be longer than or equal to 8 characters",
    },
    "property": "password",
    "target": LoginUser {},
    "value": undefined,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isDefined": "username should not be null or undefined",
      "isNotEmpty": "username should not be empty",
      "isString": "username must be a string",
      "matches": "username must match /^[\\w.-]+$/i regular expression",
      "maxLength": "username must be shorter than or equal to 30 characters",
    },
    "property": "username",
    "target": LoginUser {},
    "value": undefined,
  },
]
`;
